// chainsql-lib/dist/npm/transaction/escrow-creation.js
'use strict'; // eslint-disable-line strict

var _ = require('lodash');
var utils = require('./utils');
var _utils$common = utils.common,
    validate = _utils$common.validate,
    iso8601ToChainsqlTime = _utils$common.iso8601ToChainsqlTime,
    zxcToDrops = _utils$common.zxcToDrops,
    toChainsqldAmount = _utils$common.toChainsqldAmount;

var ValidationError = utils.common.errors.ValidationError;


function createEscrowCreationTransaction(account, payment) {
  var txJSON = {
    TransactionType: 'EscrowCreate',
    Account: account,
    Destination: payment.destination,
    Amount: toChainsqldAmount(payment.amount)
  };

  if (payment.condition !== undefined) {
    txJSON.Condition = payment.condition;
  }
  if (payment.allowCancelAfter !== undefined) {
    txJSON.CancelAfter = iso8601ToChainsqlTime(payment.allowCancelAfter);
  }
  if (payment.allowExecuteAfter !== undefined) {
    txJSON.FinishAfter = iso8601ToChainsqlTime(payment.allowExecuteAfter);
  }
  if (payment.sourceTag !== undefined) {
    txJSON.SourceTag = payment.sourceTag;
  }
  if (payment.destinationTag !== undefined) {
    txJSON.DestinationTag = payment.destinationTag;
  }
  if (payment.memos !== undefined) {
    txJSON.Memos = _.map(payment.memos, utils.convertMemo);
  }
  if (Boolean(payment.allowCancelAfter) && Boolean(payment.allowExecuteAfter) && txJSON.CancelAfter <= txJSON.FinishAfter) {
    throw new ValidationError('"CancelAfter" must be after "FinishAfter" for' + ' EscrowCreate');
  }
  return txJSON;
}

function prepareEscrowCreation(address, escrowCreation) {
  var instructions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  validate.prepareEscrowCreation({ address: address, escrowCreation: escrowCreation, instructions: instructions });
  var txJSON = createEscrowCreationTransaction(address, escrowCreation);
  return utils.prepareTransaction(txJSON, this, instructions);
}

module.exports = prepareEscrowCreation;
