// chainsql-lib/dist/npm/common/serverinfo.js
'use strict'; // eslint-disable-line strict

var _ = require('lodash');

var _require = require('./utils'),
    convertKeysFromSnakeCaseToCamelCase = _require.convertKeysFromSnakeCaseToCamelCase;

function renameKeys(object, mapping) {
  _.forEach(mapping, function (to, from) {
    object[to] = object[from];
    delete object[from];
  });
}

function getServerInfo(connection) {
  return connection.request({ command: 'server_info' }).then(function (response) {
    var info = convertKeysFromSnakeCaseToCamelCase(response.info);
    renameKeys(info, { hostid: 'hostID' });
    if (info.validatedLedger) {
      renameKeys(info.validatedLedger, {
        baseFeeZxc: 'baseFeeZXC',
        reserveBaseZxc: 'reserveBaseZXC',
        reserveIncZxc: 'reserveIncrementZXC',
        seq: 'ledgerVersion'
      });
      info.validatedLedger.baseFeeZXC = info.validatedLedger.baseFeeZXC.toString();
      info.validatedLedger.reserveBaseZXC = info.validatedLedger.reserveBaseZXC.toString();
      info.validatedLedger.reserveIncrementZXC = info.validatedLedger.reserveIncrementZXC.toString();
    }
    return info;
  });
}

function computeFeeFromServerInfo(cushion, serverInfo) {
  return (Number(serverInfo.validatedLedger.baseFeeZXC) * Number(serverInfo.loadFactor) * cushion).toString();
}

function getFee(connection, cushion) {
  return getServerInfo(connection).then(_.partial(computeFeeFromServerInfo, cushion));
}

module.exports = {
  getServerInfo: getServerInfo,
  getFee: getFee
};
