// chainsql-lib/dist/npm/ledger/accountinfo.js
'use strict'; // eslint-disable-line strict

var utils = require('./utils');
var BigNumber = require('bignumber.js');
var _utils$common = utils.common,
    validate = _utils$common.validate,
    removeUndefined = _utils$common.removeUndefined;


function formatAccountInfo(response) {
  var data = response.account_data;
  return removeUndefined({
    sequence: data.Sequence,
    zxcBalance: utils.common.dropsToZxc(data.Balance),
    ownerCount: data.OwnerCount,
    transferFeeMin: data.TransferFeeMin,
    transferFeeMax: data.TransferFeeMax,
    transferRate: data.TransferRate ? new BigNumber(data.TransferRate).shift(-9).toNumber() : undefined,
    previousInitiatedTransactionID: data.AccountTxnID,
    previousAffectingTransactionID: data.PreviousTxnID,
    previousAffectingTransactionLedgerVersion: data.PreviousTxnLgrSeq
  });
}

function getAccountInfo(address) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  validate.getAccountInfo({ address: address, options: options });

  var request = {
    command: 'account_info',
    account: address,
    ledger_index: options.ledgerVersion || 'validated'
  };

  return this.connection.request(request).then(formatAccountInfo);
}

module.exports = getAccountInfo;
